components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodTerminalService.kt

```diff
@@ -0,0 +1,382 @@
+// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package io.gitpod.jetbrains.remote
+
+import com.google.protobuf.ByteString
+import com.intellij.openapi.application.runInEdt
+import com.intellij.openapi.diagnostic.thisLogger
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.util.Key
+import com.intellij.openapi.wm.ToolWindowManager
+import com.intellij.openapi.wm.ex.ToolWindowManagerListener
+import com.jediterm.terminal.TtyConnector
+import com.jetbrains.rdserver.terminal.BackendTerminalManager
+import com.jetbrains.rdserver.terminal.BackendTtyConnector
+import io.gitpod.supervisor.api.Status.*
+import io.gitpod.supervisor.api.StatusServiceGrpc
+import io.gitpod.supervisor.api.TerminalOuterClass
+import io.gitpod.supervisor.api.TerminalServiceGrpc
+import io.grpc.StatusRuntimeException
+import io.grpc.stub.StreamObserver
+import kotlinx.coroutines.*
+import org.jetbrains.plugins.terminal.ShellTerminalWidget
+import org.jetbrains.plugins.terminal.TerminalTabState
+import org.jetbrains.plugins.terminal.TerminalToolWindowFactory
+import org.jetbrains.plugins.terminal.TerminalView
+import org.jetbrains.plugins.terminal.cloud.CloudTerminalProcess
+import org.jetbrains.plugins.terminal.cloud.CloudTerminalRunner
+import java.io.ByteArrayOutputStream
+import java.io.PipedInputStream
+import java.io.PipedOutputStream
+
+@Suppress("UnstableApiUsage")
+class GitpodTerminalService(private val project: Project) {
+    companion object {
+        val TITLE_KEY = Key.create<String>("TITLE_KEY")
+    }
+
+    private val terminalView = TerminalView.getInstance(project)
+    private val terminalServiceStub = TerminalServiceGrpc.newStub(GitpodManager.supervisorChannel)
+    private val statusServiceStub = StatusServiceGrpc.newStub(GitpodManager.supervisorChannel)
+    private val backendTerminalManager = BackendTerminalManager.getInstance(project)
+
+    init {
+        afterTerminalToolWindowGetsRegistered {
```

Comment:
I wonder would not it better to use coroutines + completable futures to have squential code instead of callbacks? 

For things like list terminal request callback looks like overkill.

Reply:
I've seen JetBrains using a lot more callbacks than completable futures, so I thought it was the best. But I agree, and will replace the callbacks by completable futures.